### PR TITLE
[ Auto Parallel ]replace softmax_with_cross_entropy with c_softmax_with_cross_entropypass

### PR DESCRIPTION
**Pull request by waliwali777**
### PR Category

 Auto Parallel 

### PR Types

Performance

### Description

基于PIR实现静半`parallel_cross_entroy` 通信算子的标记优化，在 mp 列切 + hard label 的情况下，将 `softmax_with_cross_entropy` 算子替换使用通信效率更高的 `c_softmax_with_cross_entropypass` 算子

perf in llama-7B dp1mp2pp1 with using pass
close pass: 323.2 token/card/s
open pass: 329.0 token/card/s 

Pcard-70448

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/constants.py [hunk 1]
```diff
@@ -340,7 +340,7 @@ class _DPOptimizationConfig(TypedDict, total=False):  # noqa: PYI049
 set_field_default_config(
     MP_OPTIMIZATION, "allreduce_matmul_grad_overlapping", False
 )
-
+set_field_default_config(MP_OPTIMIZATION, "replace_with_c_softmax", False)
```
Comment (liym27) [inline python/paddle/distributed/auto_parallel/constants.py]: `replace_with_c_softmax`  -> `replace_with_parallel_cross_entropy`

File: python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+logger = get_logger(logging.INFO)
+
+
+@register_pass("replace_with_parallel_cross_entropy")
+class AutoParallelReplaceWithParallelCrossEntropyPass(PassBase):
+    def __init__(self):
+        super().__init__()
+        self._in_pir_mode = paddle.base.framework.get_flags(
+            "FLAGS_enable_pir_api"
+        )["FLAGS_enable_pir_api"]
```
Comment (liym27) [inline python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py]: self._in_pir_mode is not used

File: python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                    # meed split along the column.
+                    placement1 = operand1.placements
+                    if not placement1[1].is_shard():
+                        return
+
+                    ignore_index = op.attrs()["ignore_index"]
+                    paddle.pir.set_insertion_point(op)
+                    softmax, loss = paddle._C_ops.c_softmax_with_cross_entropy(
+                        operand1, operand2, ignore_index, ring_id, rank, nranks
+                    )
```
Comment (liym27) [inline python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py:80]: check the spmd of other operators after `c_softmax_with_cross_entropy`

**Review by liym27 — COMMENTED** [review]
add a todo: add llama testing with this pass

File: python/paddle/distributed/auto_parallel/constants.py (hunk 1 above)
Comment (waliwali777) [reply python/paddle/distributed/auto_parallel/constants.py]: done

File: python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py (hunk 2 above)
Comment (waliwali777) [reply python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py]: removed

File: python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py (hunk 3 above)
Comment (waliwali777) [reply python/paddle/distributed/passes/auto_parallel_replace_with_parallel_cross_entropy.py:80]: Added `_check_user` func in pass

**Review by XieYunshen — APPROVED** [review]

**Review by liym27 — APPROVED** [review]
LGTM
